bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerOpenOp.java

```diff
@@ -112,12 +112,10 @@ public void initiate() {
          * Asynchronously read the ledger metadata node.
          */
         bk.getLedgerManager().readLedgerMetadata(ledgerId)
-            .whenComplete((metadata, exception) -> {
-                    if (exception != null) {
-                        openComplete(BKException.getExceptionCode(exception), null);
-                    } else {
-                        openWithMetadata(metadata);
-                    }
+                .thenAcceptAsync(this::openWithMetadata, bk.getScheduler().chooseThread(ledgerId))
+                .exceptionally(exception -> {
+                    openComplete(BKException.getExceptionCode(exception), null);
```

Comment:
After thinking more...
We have to execute also this callback in a separate thread.
I suggest to use whenCompleteAsync

Reply:
Then, we need to create a new executor here if bk is closed.

Reply:
Then, we need to create new executor here  for exception case.

Reply:
```
.whenCompleteAsyncAsync((m, exception) -> {
if (exception != null) {
 openComplete(BKException.getExceptionCode(exception), null);
}  else {
this.openWithMetadata(m);
}
}, bk.getScheduler().chooseThread(ledgerId))
```

Reply:
`org.apache.bookkeeper.client.BookKeeperCloseTest#testBookKeeperAdmin`  This test will be stuck, if change made

Reply:
This happened if bk is closed, we cannot go inside `whenCompleteAsyncAsync` , because bk's scheduler is shutdown.

Reply:
FYI @eolivelli

Reply:
So when the BK closed the treadpool is shutdown and the callback isn't executed.
This looks like a problem that can happen on every callback.

Let's keep this patch as it is here but please open a ticket to track this bad problem.

If the BK client is closed then probably the application is shutting down  and it is a critical point in the life cycle